sqlmesh/core/macros.py

```diff
@@ -349,6 +354,65 @@ def engine_adapter(self) -> EngineAdapter:
             )
         return self.locals["engine_adapter"]
 
+    def _coerce(self, expr: exp.Expression, typ: t.Type) -> t.Any:
+        """Coerces the given expression to the specified type on a best-effort basis."""
+        try:
+            if typ is None:
+                return expr
+            base = t.get_origin(typ) or typ
+            if isinstance(expr, base):
+                return expr
+            if t.get_origin(typ) is t.Union:
+                for t_ in t.get_args(typ):
+                    try:
+                        return self._coerce(expr, t_)
+                    except Exception:
+                        pass
+                raise SQLMeshError(f"Failed to coerce expression '{expr}' to type '{typ}'.")
+            if issubclass(base, exp.Expression):
+                d = Dialect.get_or_raise(self.dialect)
+                into = base if base in d.parser().EXPRESSION_PARSERS else None
+                coerced = parse_one(
+                    expr.this if isinstance(expr, exp.Literal) else expr.sql(), into=into
```

Comment:
this doesn't look right to me anymore. i don't think you should call parse_one(sql()) if there's no into and not a literal

Reply:
I'll take another look. My aim was to be super relaxed and make a best-effort in all cases since we will always fallback to the originally parsed type. There is an `isinstance` after this to make sure we don't give back anything jank and just give back the original if we didnt get it right. 
But I think what you say on the lit and into None combo being its own case...

Reply:
what case are you expecting with no into and expr.sql()?

Reply:
yeah I see what you mean, itd be a noop at best. fixed